### PR TITLE
Clip planes support added

### PR DESCRIPTION
**Pull request by bilgili**

### CODE REVIEW

File: livre/core/render/ClipPlanes.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/* Copyright (c) 2011-2015, EPFL/Blue Brain Project
```
Comment (rdumusc) [inline livre/core/render/ClipPlanes.cpp]: wrong date ;-)

File: livre/core/render/ClipPlanes.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+void ClipPlanes::clear()
+{
+    getPlanes().clear();
+}
+
+void ClipPlanes::reset()
+{
+    getPlanes().clear();
```
Comment (rdumusc) [inline livre/core/render/ClipPlanes.cpp]: just clear()

File: livre/eq/render/RayCastRenderer.cpp [hunk 3]
```diff
@@ -503,7 +529,6 @@ struct RayCastRenderer::Impl
     const VolumeInformation& _volInfo;
     GLuint _posVBO;
     GLuint _quadVBO;
-    GLint _drawBuffer;
```
Comment (rdumusc) [inline livre/eq/render/RayCastRenderer.cpp]: draw buffer?

File: livre/eq/render/shaders/fragRayCast.glsl [hunk 4]
```diff
@@ -40,6 +40,9 @@ uniform int nSamplesPerPixel;
 uniform float shininess;
 uniform int refLevel;
 
+uniform int nClipPlanes;
+uniform vec4 clipPlanes[ 8 ];
```
Comment (rdumusc) [inline livre/eq/render/shaders/fragRayCast.glsl]: 8 -> 6

File: livre/eq/settings/RenderSettings.h [hunk 5]
```diff
@@ -56,14 +58,26 @@ class RenderSettings : public co::Serializable
...
+        { return _transferFunction; }
+
+    /**
+     * @brief Sets the clip planes.
+     * @param clipPlanes the clip planes
+     */
+    void setClipPlanes( const ClipPlanes& clipPlanes );
+
+    /**
+     * @return Returns the transfer function.
```
Comment (rdumusc) [inline livre/eq/settings/RenderSettings.h]: Returns the ClipPlanes

File: livre/eq/zeroeq/communicator.cpp [hunk 6]
```diff
@@ -147,17 +147,15 @@ class Communicator::Impl
 
     void _setupRequests()
     {
-        _requests[ _frame.getTypeIdentifier() ] = [&]
-            { return publishFrame(); };
+        _requests[ _frame.getTypeIdentifier() ] = [&]{ return publishFrame(); };
         _requests[ _getFrameData().getVRParameters().getTypeIdentifier( )] = [&]
-            { return _publisher.publish( _getFrameData().getVRParameters( )); };
+            {   return _publisher.publish( _getFrameData().getVRParameters( )); };
```
Comment (rdumusc) [inline livre/eq/zeroeq/communicator.cpp]: style

File: livre/eq/zeroeq/communicator.cpp [hunk 7]
```diff
@@ -147,17 +147,15 @@ class Communicator::Impl
...
         _requests[ _getFrameData().getCameraSettings().getTypeIdentifier( )] = [&]
-            { return _publisher.publish( _getFrameData().getCameraSettings());};
+            {   return _publisher.publish( _getFrameData().getCameraSettings()); };
```
Comment (rdumusc) [inline livre/eq/zeroeq/communicator.cpp]: style

File: livre/eq/zeroeq/communicator.cpp [hunk 8]
```diff
@@ -147,17 +147,15 @@ class Communicator::Impl
...
         _requests[ _getRenderSettings().getTransferFunction().getTypeIdentifier( )] = [&]
-            {
-                return _publisher.publish(
-                    _getRenderSettings().getTransferFunction( ));
-            };
+            { return _publisher.publish( _getRenderSettings().getTransferFunction( )); };
+        _requests[ _getRenderSettings().getClipPlanes().getTypeIdentifier( )] = [&]
+            {   return _publisher.publish( _getRenderSettings().getClipPlanes( )); };
```
Comment (rdumusc) [inline livre/eq/zeroeq/communicator.cpp]: style

File: livre/lib/pipeline/RenderFilter.cpp [hunk 9]
```diff
@@ -62,7 +62,8 @@ struct RenderFilter::Impl
         {
             { "CacheObjects" ,getType< ConstCacheObjects >() },
             { "Frustum" ,getType< Frustum >() },
-            { "Viewport" ,getType< PixelViewport >() }
+            { "Viewport" ,getType< PixelViewport >() },
+            { "ClipPlanes" ,getType< ClipPlanes >() }
```
Comment (rdumusc) [inline livre/lib/pipeline/RenderFilter.cpp]: style ,

File: tests/core/clipPlanes.cpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
+/* Copyright (c) 2015, EPFL/Blue Brain Project
```
Comment (rdumusc) [inline tests/core/clipPlanes.cpp]: 2016

File: tests/core/clipPlanes.cpp [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+ *                     Jafet.VillafrancaDiaz@epfl.ch
```
Comment (rdumusc) [inline tests/core/clipPlanes.cpp]: hello Jafet :-)

File: apps/livreGUI/clipPlanesController/ClipPlanesController.cpp [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        : _parent( parent )
+        , _controller( controller )
+    {
+        _ui.setupUi( parent );
+
+        QWidget *slidersWidget = _ui.clipPlanesWidget;
+        slidersWidget->setEnabled( false);
+
+        parent->connect( parent,
+            static_cast<void (ClipPlanesController::*)()>(&ClipPlanesController::clipPlanesReceived),
```
Comment (rdumusc) [inline apps/livreGUI/clipPlanesController/ClipPlanesController.cpp]: Why all these static cast? I don't remember having to do this

File: apps/livreGUI/clipPlanesController/ClipPlanesController.cpp [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+                _clipPlanes.getPlanes()[ i ].setD( i % 2 == 0 ? spinBoxValue : -spinBoxValue );
+                changed = true;
+            }
+            ++it;
+            ++i;
+        }
+
+        if( changed )
+           _controller.publish( _clipPlanes );
+    }
```
Comment (rdumusc) [inline apps/livreGUI/clipPlanesController/ClipPlanesController.cpp]: missing \n

File: apps/livreGUI/clipPlanesController/ClipPlanesController.cpp (hunk 12 above)
Comment (bilgili) [inline apps/livreGUI/clipPlanesController/ClipPlanesController.cpp]: I did similar to renderParametersController. It did work so I did not change it, but I can try removing

**Comment by rdumusc** [comment]
OK +1, squash commits before merging

**Pull request merged**